src/mds/MDBalancer.cc

```diff
@@ -1061,17 +1069,33 @@ void MDBalancer::hit_dir(utime_t now, CDir *dir, int type, int who, double amoun
 	mds->mdsmap->allows_dirfrags() &&
 	(dir->should_split() ||
 	 (v > g_conf->mds_bal_split_rd && type == META_POP_IRD) ||
-	 (v > g_conf->mds_bal_split_wr && type == META_POP_IWR)) &&
-	split_queue.count(dir->dirfrag()) == 0) {
-      dout(10) << "hit_dir " << type << " pop is " << v << ", putting in split_queue: " << *dir << dendl;
-      split_queue.insert(dir->dirfrag());
+	 (v > g_conf->mds_bal_split_wr && type == META_POP_IWR))) {
+      dout(10) << "hit_dir " << type << " pop is " << v << ", putting in split_pending: " << *dir << dendl;
+      if (split_pending.count(dir->dirfrag()) == 0) {
+        queue_split(dir, false);
+      } else {
+        // If the dir size is too far ahead of bal_split_size, then do a
+        // mds->queue_waiter to action the split at the end of the current
+        // request instead of waiting for the timer.
+        // Count null dentries in effective size to take account of insertions
+        // that haven't committed yet.
+        auto effective_size = dir->get_frag_size() + dir->get_num_head_null();
```

Comment:
Can't we get null dentries from actually deleting things, or doing lookups on incomplete CDirs?